### DISCUSSION TITLE
Adding testcase dynamically before start of testing

### DISCUSSION THREAD

**Discussion by Stefanhg** [pytest-dev/pytest#12221] [Q&A]
Hello,
I want to add a custom testcase to the `pytest_collection_modifyitems` items list before my test session starts depending on a variable set in the pytest `config`.

I played around with different functions and got stuck at this `from_parent` function which I haven't had any success with.
I do have a "work-around" which is just including it with `-k` and give a warning if variable is set in config and the testcase is missing in collector but what's the fun in that!

My goal is to:
1. Add the testcase to the testsuite, 
2. To have it as the first testcase in the testsuite.

Simplified example of what i want to do, it is kind of hard to show when you don't know how to implement.
I am unsure if it is even possible to do this way or it has to go though the collector.
```
# test_mytests.py
def test_first_test(): pass

def test_dynamic_test(): pass
```
```
# conftest.py
def pytest_collection_modifyitems(session, config, items):
    item = add_testcase_to_current_testsuite(obj=test_dynamic_test, name="test_dynamic_test")
    items.insert(0, item)
```

**Comment by portyu9** [comment]
Hi @Stefanhg 

If `test_dynamic_test` is a normal Python test function, I would **collect it normally** and use `pytest_collection_modifyitems` only to decide whether it stays in the run and where it is ordered.

`pytest_collection_modifyitems` runs **after collection has completed**, so manufacturing a new `pytest.Function` there with `from_parent()` bypasses parts of the normal collection lifecycle.

For example, mark the test:

```python
import pytest


@pytest.mark.preflight
def test_dynamic_test():
    pass
```

Then conditionally keep/order it:

```python
def pytest_collection_modifyitems(config, items):
    preflight = [
        item for item in items
        if item.get_closest_marker("preflight")
    ]
    other = [
        item for item in items
        if not item.get_closest_marker("preflight")
    ]

    if config.getoption("--run-preflight"):
        items[:] = preflight + other
    else:
        config.hook.pytest_deselected(items=preflight)
        items[:] = other
```

This gives you the desired behavior without manually constructing pytest internals:

1. the test runs only when required,
2. it is first when enabled,
3. normal fixture/marker/plugin collection processing still happens.

If the testcase genuinely does not exist until runtime data is examined, then yes — create it **through collection**, using an appropriate collector/collection hook and `from_parent()`, rather than inserting it in `pytest_collection_modifyitems`.

Please mark this answer as accepted if it helped, thank you.
